### PR TITLE
[flang][OpenMP] Implement flexible OpenMP clause representation

### PR DESCRIPTION
**Pull request by kparzysz**
The new set of classes representing OpenMP classes mimics the contents of parser::OmpClause, but differs in a few aspects:
- it can be easily created, copied, etc.
- is based on semantics::SomeExpr instead of parser objects.

This set of classes is geared towards a language-agnostic representation of OpenMP clauses.  While the class members are still based on flang's `parser::OmpClause`, the classes themselves are actually C++ templates parameterized with types essential to represent necessary information.  The two parameters are
- `IdType`: the type that can represent object's identity (for flang it will be `semantics::Symbol *`),
- `ExprType`: the type that can represent expressions, arithmetic and object references (`semantics::MaybeExpr` in flang).

The templates are defined in a namespace `tomp` (to distinguish it from `omp`).

This patch introduces file "Clauses.cpp", which contains instantiations of all of the templates for flang.  The instantiations are members of namespace `omp`, and include an all-encompassing variant class `omp::Clause`, which is the analog of `parser::OmpClause`.
The class `OmpObject` is represented by `omp::Object`, which contains the symbol associated with the object, and `semantics::MaybeExpr` representing the designator for the symbol reference.  For each specific clause in the variant `parser::OmpClause`, there exists a `make` function that will generate the corresponding `omp::Clause` from it.  The intent was to use the make functions as variant visitors.  The creation of a clause instance would then follow the pattern:
```
omp::Clause clause = std::visit([](auto &&s) { return make(s, semaCtx); }, parserClause.u);
```

If a new clause `foo` is added in the future, then:
- a new template `tomp::FooT` representing the clause needs to be added to ClauseT.h,
- a new instance needs to be created in flang, this can be as simple as `using Foo = tomp::FooT<...>`,
- a new make function needs to be implemented to create object of class Foo from `parser::OmpClause::Foo`.

This patch only introduces the new classes, they are not yet used anywhere.

[Clause representation 1/6]

### CODE REVIEW

**Comment by kparzysz** [comment]
Added reuse of enum definitions from Fortran::parser, as suggested by @skatrak in the community call.

**Review by skatrak — COMMENTED** [review]
Thank you Krzysztof for this work. I added a few comments inline, but they are rather minor nits and suggestions.

I see that this approach requires representing the complete list of clauses, and that's the reason for the size of this patch. However, these are very much tied to the flang PFT, so I wonder whether it would be feasible to get this to generically work for clang as well and make it part of llvm/Frontend/OpenMP, as it was discussed in yesterday's community call.

For example, the reduction clause is defined here and in the PFT as a `ReductionOperator` and an `ObjectList`, and the `ReductionOperator` is made of a `DefinedOperator` or a `ProcedureDesignator`. In [clang](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/AST/OpenMPClause.h#L3204), the corresponding structure inherits from `OMPVarListClause` and contains a `DeclarationNameInfo`, among others. So I can foresee some issues creating `make()` functions for these structures based on the clang AST, since representations are quite different. Also, the various traits are only used by flang's PFT visitors and the main `Object` class here uses `Fortran::semantics` types, which I'm not sure whether it could be easily templated to contain either flang or clang symbols/expressions.

So I'm not sure whether the list of structures here could be made somewhat more language agnostic and based off of the spec rather than flang's PFT if there's a chance that this approach could be shared by clang and flang.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:150]: I usually run `clang-format` from the root directory of the LLVM project, and it puts the following type declaration on the same line with the template, if it fits.  Apparently the lowering code uses MLIR's clang-format settings (which have `AlwaysBreakTemplateDeclarations: Yes`).  The MLIR's format settings are in the mlir subdirectory and are not applied here by default.  The CI format checker expects the MLIR rules, and if I just clang-format myself, it will fail

I'm not sure what exactly is happening here with the formatting, but I put the trailing comment to keep the template header separate even with my settings.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:176]: It used to be, but some classes that ended up in `List` were too large for SmallVector's default size estimation (200-something bytes), so I converted it to `std::vector`.
To elaborate: `SmallVector` takes two template arguments: type and initial size.  The initial size used to be mandatory, but at some point it was made optional, and the initial size was computed based on the type.  If the type exceeds some specific size, the estimation will no longer work, and the code will fail to compile.  Instead of adding a specific number, I just changed it to `std::vector`.  It can be changed back though.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:372]: There shouldn't be any copies here---the object should be created in-place.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:355]: It's an omission.  Both of these macros are "global" in the sense that they are defined in some header and used elsewhere.  Since I redefine them, I don't want my redefinitions to leak out and cause confusion.

The TableGen-generated code should not be using these macros directly. It's an easy fix, and would solve this issue.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:298]: The semantic analyzer will visit `parser::Expr` nodes in the AST, create `evaluate::Expr` objects for them, and store the evaluator objects in a "secret" field in the parser expression.  The function `GetExpr` retrieves the stored value.  The problem here was that I wanted an `evaluate::Expr` that represents a designator, for which the precomputed `evaluate::Expr` was not available.  That's why I added code that builds the evaluator expressions when needed.

**Review by kparzysz — COMMENTED** [review]
Thanks for the comments.  I replied with some rationale (or background) for some of the decisions you commented on.

I'm working on separating this code from Fortran's AST right now.

File: flang/lib/Lower/OpenMP.cpp
Comment (clementval) [reply flang/lib/Lower/OpenMP.cpp:355]: >The TableGen-generated code should not be using these macros directly.
Why? 
`GEN_FLANG_CLAUSE_PARSER_CLASSES ` is generated for the parse-tree and just not meant to be included here.

Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:355]: Because it forces a specific interface between the autogenerated .inc file and its uses in the C++ sources.  TableGen-generated files that are "enumerative" in nature shouldn't need to be aware of the context where they're used.  Things that have a single application today may have multiple tomorrow.

Comment (clementval) [reply flang/lib/Lower/OpenMP.cpp:355]: Yes and this is indeed induced by the name of the including def. There is no harm to have another entry point in the DirectiveEmitter that produces a more generic list that is not meant for the parser nodes. This would also remove the need to use macro that are used specifically in flang parse-tree.

Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:355]: What TG generates here is a list of pairs (class-name, contents).  In some cases we may need both, in other cases we may only need the class name.  Conceptually, both are different pieces of the same root information.  We can indeed generate more information in TG, but we wouldn't need to if the information was done generically from the start.

**Comment by kparzysz** [comment]
> So I'm not sure whether the list of structures here could be made somewhat more language agnostic and based off of the spec rather than flang's PFT if there's a chance that this approach could be shared by clang and flang.

This is a very good point.  Yes, that would be better, and that's the next step here, but it would be good to have something working before that, so the changes are more incremental.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:150]: Got rid of the trailing comments.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:219]: Changed it to return `void`.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:322]: Done

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:340]: It's actually illegal to redefine a macro with a different definition, although usually you'd only get a warning.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp:179]: Renamed to `SymbolWithDesignator`.

**Comment by kparzysz** [comment]
Ping.  Are there still concerns about this?

**Comment by kiranchandramohan** [comment]
Could you expand the summary to briefly mention the contents of the patch?
What is the mapping from parser::Clause to omp::Clause? What is omp::ClauseT? What are the various `make` functions?
In future, if a new clause is added what are the changes that have to be made for the new representation? 
Would it make sense to construct this during PFT creation?

**Comment by kparzysz** [comment]
Thanks Kiran.  I expanded the PR description with additional information.

The creation of omp::Clause from parser::OmpClause can be done with a single function call:
```
makeClause(parser::OmpClause, semaCtx) -> omp::Clause
```
so the complexity of creating one is relatively small.  If we created them during PTF creation, we'd need to store them somewhere (in OmpClause itself?).  Replacing OmpClause with omp::Clause would require potentially invasive changes in the parser/semantics analyzer, plus it would deviate from the representations of other PFT nodes (specifically SomeExpr/MaybeExpr is not explicitly used in PFT, except in parser::Expr).

I'm not sure if this answers your question though.  Let me know if you'd like more information.

**Review by skatrak — COMMENTED** [review]
Thank you Krzysztof for addressing my comments, I mostly just have some nits left but it looks good in general and I agree with the approach of defining these Fortran-specific structures first and then transition to something shared with Clang later.

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp]: Could `SymDsgExtractor` be renamed to `SymbolWithDesignatorExtractor` as well? I prefer it that way, though it's a bit verbose. I'll leave it up to you if you prefer to keep it as it is.

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp]: Ah yes, I'm aware of this limitation. Do you think it would be possible to store const references to those larger objects rather than full copies? I'm guessing these might be Symbol or Expression-related types, which I've seen trigger this sort of thing before. If we can't get around copying these big structures, then either keeping `std::vector` or `llvm::SmallVector<T, 0>` would be necessary.

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp]: Can it be named `makeObjectFn` or `makeObjectCB` instead? That way it's clearer that it's intended to return a callback. The "F" on its own didn't quite make it clear enough for me.

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp]: Even if the compiler is ultimately able to avoid the copy constructor, `std::visit` here already returns a `DefinedOperator` object. Below, in `makeRedOp` and `make` functions for `Depend` and `Linear` it's returned directly, as I'm suggesting here. I still think it's worthwhile to make that change even if only for uniformity reasons.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: We're creating these objects, so we need to manage storage for them.  If we wanted to store references, we'd have to have another container that actually holds the referenced objects.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Done

File: flang/lib/Lower/OpenMP/ClauseT.h
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/ClauseT.h:57]: I'd prefer not---this would create too much clutter.  I specifically avoided spelling it out everywhere, except where the parameters are introduced.

File: flang/lib/Lower/OpenMP/Clauses.h
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.h:96]: Done

File: flang/lib/Lower/OpenMP/Clauses.h
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.h:127]: The empty clauses are specialized automatically in the macro+include above.  I can delete `AtomicDefaultMemOrder` if you want, but then we'd lose completeness.  I can add a comment here explaning the "requires" case.  Let me know what you'd prefer.

This may all be moot, because the generic variant visitors may require specializations and "make" functions for every alternative in the variant.  I'll try to see what happens if I remove this one specialization.

Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.h:127]: Yep, it failed to compile.

File: flang/lib/Lower/OpenMP/Clauses.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.cpp:56]: The `detail` namespace is a convention used to avoid putting helper things into the same namespace as the visible API.  This mostly applies to header files in cases of helper templates (which cannot be hidden in the .cpp file).

I propose that we keep it here for now, we can always move them elsewhere later on.

File: flang/lib/Lower/OpenMP/Clauses.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.cpp:281]: Done

File: flang/lib/Lower/OpenMP/Clauses.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.cpp:214]: Done

File: flang/lib/Lower/OpenMP/Clauses.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.cpp:230]: Done

File: flang/lib/Lower/OpenMP/Clauses.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.cpp:242]: Done

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp]: I was hoping we only needed to create clause structures and we were able to keep references to original symbols and expressions, but we're creating new expressions here then so you're right that it doesn't make sense to try to store them as references. I think for cases like this `llvm::SmallVector<T, 0>` is preferred over `std::vector`, so I'd still suggest making that small change.

File: flang/lib/Lower/OpenMP/ClauseT.h
Comment (skatrak) [reply flang/lib/Lower/OpenMP/ClauseT.h:57]: We're going to have to agree to disagree on this one, but I won't block approving this patch based on this minor thing which is largely just a matter of preference 😄. Maybe someone else can chime in to make a decision.

File: flang/lib/Lower/OpenMP/Clauses.h
Comment (skatrak) [reply flang/lib/Lower/OpenMP/Clauses.h:134]: I see, it needs to be there, so there it shall remain. Thanks for checking, though.

It may be nice to add a comment to the `make` function for `AtomicDefaultMemOrder` to basically say "this won't be called because requires clauses are processed prior to lowering", to avoid potential confusion.

File: flang/lib/Lower/OpenMP/Clauses.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/Clauses.cpp:56]: You're right, I assumed I was looking at a header file when I made the comment. I'd still rather see these functions outside the struct, but it's ok if you prefer leaving them there.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Done

File: flang/lib/Lower/OpenMP/Clauses.h
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Clauses.h:134]: Added a comment in .cpp, and in .h where AtomicDefaultMemOrder is instantiated (in the previous patch).

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Done

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: I think leaving this as `ContainerTy` makes it more generic.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Renamed to `makeObjectFn`.

File: flang/lib/Lower/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Done

**Review by skatrak — APPROVED** [review]
Thank you Krzysztof for all the work (and for your patience with my nits!). I think this is good to go, but since it's a big change and there are a few other PRs coming right after, I'd suggest waiting for @kiranchandramohan's approval as well before merging.

**Comment by kparzysz** [comment]
Windows build: `fatal error C1060: compiler is out of heap space`

**Comment by kparzysz** [comment]
Hi @kiranchandramohan, are you ok with merging this?

**Comment by kiranchandramohan** [comment]
> Hi @kiranchandramohan, are you ok with merging this?

Yes. Nice work! And thanks @skatrak for the detailed review.

**Comment by kparzysz** [comment]
Again seeing a bunch of `fatal error C1060: compiler is out of heap space` in the Windows build.  The builds have passed in the past (without major changes since then), so I'm merging it.

**Comment by kparzysz** [comment]
That's fixed in https://github.com/llvm/llvm-project/commit/6479218932117ab36cba578bb458ee7c0fb4bf0a

Plus NDEBUG issue, fixed in https://github.com/llvm/llvm-project/commit/5a77bdc3e7d60c1724054d09a7dc7c67fb58ea9a

**Pull request merged**
